tsl/src/compression/api.c

```diff
@@ -1075,128 +1059,682 @@ tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+	tuplesort_performsort(uncompressed_rows_sortstate);
 }
 
 static void
-fetch_matching_uncompressed_chunk_into_tuplesort(Tuplesortstate *segment_tuplesortstate,
-												 int nsegmentby_cols,
-												 Relation uncompressed_chunk_rel,
-												 CompressedSegmentInfo **current_segment)
+fetch_segment_metadata(RowDecompressor *decompressor, HeapTuple compressed_tuple,
+					   Datum *metacount_val, Datum *sqlnum_val)
 {
-	TableScanDesc heapScan;
-	HeapTuple uncompressed_tuple;
-	TupleDesc uncompressed_rel_tupdesc = RelationGetDescr(uncompressed_chunk_rel);
-	int index = 0;
-	int nsegbycols_nonnull = 0;
-	Bitmapset *null_segbycols = NULL;
+	Datum *values = palloc0(sizeof(Datum) * decompressor->in_desc->natts);
+	bool *nulls = palloc0(sizeof(bool) * decompressor->in_desc->natts);
+	heap_deform_tuple(compressed_tuple, decompressor->in_desc, values, nulls);
+
+	if (sqlnum_val)
+	{
+		/* fetch sequence number of current compressed tuple */
+		AttrNumber seq_num =
+			get_attnum(decompressor->in_rel->rd_id, COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME);
+		*sqlnum_val = values[seq_num - 1];
+	}
+	if (metacount_val)
+	{
+		/* fetch number of rows present in current segment of compressed tuple */
+		AttrNumber meta_count =
+			get_attnum(decompressor->in_rel->rd_id, COMPRESSION_COLUMN_METADATA_COUNT_NAME);
+		*metacount_val = values[meta_count - 1];
+	}
+	pfree(values);
+	pfree(nulls);
+}
+
+static int
+decompress_in_tuplestore(RowDecompressor *decompressor, HeapTuple compressed_tuple,
+						 Tuplesortstate *merge_tuplestore)
+{
+	Assert(HeapTupleIsValid(compressed_tuple));
+	heap_deform_tuple(compressed_tuple,
+					  decompressor->in_desc,
+					  decompressor->compressed_datums,
+					  decompressor->compressed_is_nulls);
+	/* fetch sequence number of current compressed tuple */
+	AttrNumber seq_num =
+		get_attnum(decompressor->in_rel->rd_id, COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME);
+	/* decompress found tuple and delete from compressed chunk */
+	row_decompressor_decompress_row(decompressor, merge_tuplestore);
+	TM_FailureData tmfd;
+	TM_Result result pg_attribute_unused();
+	result = table_tuple_delete(decompressor->in_rel,
+								&compressed_tuple->t_self,
+								decompressor->mycid,
+								GetTransactionSnapshot(),
+								InvalidSnapshot,
+								true,
+								&tmfd,
+								false);
+	Assert(result == TM_Ok);
+	/* make changes visible */
+	CommandCounterIncrement();
+	return decompressor->compressed_datums[seq_num - 1];
+}
 
-	for (int seg_col = 0; seg_col < nsegmentby_cols; seg_col++)
+static int
+fetch_compressed_tuples_per_segment(RowDecompressor *decompressor, Chunk *uncompressed_chunk,
+									CompressedSegmentInfo **current_segment, int sequence_num,
+									int nsegmentby_cols, Tuplesortstate *tuplestore)
+{
+	int segments_decompressed = 0;
+	/* build scankeys for segmentby and sequence number columns */
+	ScanKeyData *scankeys = palloc0((nsegmentby_cols + 1) * sizeof(ScanKeyData));
```

Comment:
You might want to release this memory after you're done with it. Doesn't seem like it need to remain after the scan.

Reply:
Fixed.